neqo-transport/src/send_stream.rs

```diff
@@ -1063,61 +1115,283 @@ impl ::std::fmt::Display for SendStream {
+    // a single stream at a given sendorder.
+
+    // These both store stream_ids, which need to be looked up in 'map'.
+    // This avoids the complexity of trying to hold references to the
+    // Streams which are owned by the IndexMap.
+    sendordered: BTreeMap<SendOrder, OrderGroup>,
+    regular: OrderGroup, // streams with no SendOrder set, sorted in stream_id order
+}
 
 impl SendStreams {
     pub fn get(&self, id: StreamId) -> Res<&SendStream> {
-        self.0.get(&id).ok_or(Error::InvalidStreamId)
+        self.map.get(&id).ok_or(Error::InvalidStreamId)
     }
 
     pub fn get_mut(&mut self, id: StreamId) -> Res<&mut SendStream> {
-        self.0.get_mut(&id).ok_or(Error::InvalidStreamId)
+        self.map.get_mut(&id).ok_or(Error::InvalidStreamId)
     }
 
     pub fn exists(&self, id: StreamId) -> bool {
-        self.0.contains_key(&id)
+        self.map.contains_key(&id)
     }
 
     pub fn insert(&mut self, id: StreamId, stream: SendStream) {
-        self.0.insert(id, stream);
+        self.map.insert(id, stream);
+    }
+
+    pub fn set_sendorder(&mut self, stream_id: StreamId, sendorder: Option<SendOrder>) -> Res<()> {
+        if let Some(stream) = self.map.get_mut(&stream_id) {
+            // don't grab stream here; causes borrow errors
+            let old_sendorder = stream.sendorder();
+            if old_sendorder != sendorder {
+                // we have to remove it from the list it was in, and reinsert it with the new
+                // sendorder key
+                let mut group = if let Some(old) = old_sendorder {
+                    self.sendordered.get_mut(&old).unwrap()
+                } else {
+                    &mut self.regular
+                };
+                group.remove_streamid(&stream_id);
+                self.get_mut(stream_id).unwrap().set_sendorder(sendorder);
+                if let Some(order) = sendorder {
+                    group = self.sendordered.entry(order).or_default();
+                } else {
+                    group = &mut self.regular;
+                }
+                group.insert_streamid(&stream_id);
+                qtrace!(
+                    "ordering of stream_ids: {:?}",
+                    self.sendordered.values().collect::<Vec::<_>>()
+                );
+            }
+            Ok(())
+        } else {
+            Err(Error::InvalidStreamId)
+        }
+    }
+
+    pub fn set_fairness(&mut self, stream_id: StreamId, make_fair: bool) -> Res<()> {
+        if let Some(stream) = self.map.get_mut(&stream_id) {
+            let old_fair = stream.fair;
+            stream.set_fairness(make_fair);
+            if !old_fair && make_fair {
+                // move to the regular OrderGroup.  We must be called before SendOrder can be set
+                match stream.sendorder {
+                    None => {},
+                    Some(_) => panic!("Stream had sendOrder set before fairness!"),
```

Comment:
I am not a fan of strict ordering of calls like this.
Maybe add another parameter `fair` in `set_sendorder` and remove this one.
What do you think?

Reply:
Modified set_sendorder to call set_fairness if it's not already fair, and removed this  match/panic